[Cocoa] Align <datalist> dropdown keyboard interactions with platform conventions

pxlcoder opened the pull request:
#### e638aadaf01611f07793a35ef8dc7b77e7ad4dec
<pre>
[Cocoa] Align &lt;datalist&gt; dropdown keyboard interactions with platform conventions
<a href="https://bugs.webkit.org/show_bug.cgi?id=286045">https://bugs.webkit.org/show_bug.cgi?id=286045</a>
<a href="https://rdar.apple.com/143012287">rdar://143012287</a>

Reviewed by Abrar Rahman Protyasha.

- Up and down arrow keys should not cycle through options
- Escape key should dismiss the dropdown

* LayoutTests/fast/forms/datalist/datalist-hide-using-escape-key-expected.txt: Added.
* LayoutTests/fast/forms/datalist/datalist-hide-using-escape-key.html: Added.
* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options-on-keydown.html:
* LayoutTests/fast/forms/datalist/datalist-textinput-keydown-expected.txt:
* LayoutTests/fast/forms/datalist/datalist-textinput-keydown.html:
* LayoutTests/fast/forms/datalist/datalist-textinput-suggestions-order.html:
* LayoutTests/platform/gtk/fast/forms/datalist/datalist-textinput-keydown-expected.txt: Added.
* LayoutTests/platform/ipad/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.waitForDataListSuggestionsToChangeVisibility):
(window.UIHelper.isShowingDateTimePicker): Deleted.
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::handleKeydownEvent):
* Source/WebKit/UIProcess/mac/WebDataListSuggestionsDropdownMac.mm:
(-[WKDataListSuggestionsController moveSelectionByDirection:]):
* Source/WebKit/WebProcess/WebCoreSupport/WebDataListSuggestionPicker.cpp:
(WebKit::WebDataListSuggestionPicker::handleKeydownWithIdentifier):

Canonical link: <a href="https://commits.webkit.org/289067@main">https://commits.webkit.org/289067@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e360f11413f4557df64d26f4c7960145886d1687

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4759 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39427 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90177 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36085 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87116 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4849 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12735 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66175 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23987 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88074 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3712 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77272 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46441 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3597 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31503 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35158 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74404 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32314 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91560 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12373 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9053 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74669 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12602 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73085 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73786 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18171 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16621 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4387 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13284 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12321 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17793 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12157 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15652 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13903 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/html/TextFieldInputType.cpp:
```diff
@@ -209,7 +209,7 @@ auto TextFieldInputType::handleKeydownEvent(KeyboardEvent& event) -> ShouldCallB
 #if ENABLE(DATALIST_ELEMENT)
     const String& key = event.keyIdentifier();
-    if (RefPtr suggestionPicker = m_suggestionPicker; suggestionPicker && (key == "Enter"_s || key == "Up"_s || key == "Down"_s)) {
+    if (RefPtr suggestionPicker = m_suggestionPicker; suggestionPicker && (key == "Enter"_s || key == "Up"_s || key == "Down"_s || key == "U+001B"_s)) {
```
Can you add a comment saying which key this is

pxlcoder replied:
hmm we have this in quite a few places with no comment

pascoej replied:
I think this might better be should be translated to "Escape" in `PlatformEventFactory*`

pxlcoder replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=286147 for that.

aprotyas commented on LayoutTests/fast/forms/datalist/datalist-hide-using-escape-key.html:
```diff
@@ -0,0 +1,68 @@
+            continue;
+        resolve();
+    });
+}
```
Maybe push this into `UIHelper`?

pxlcoder committed:
[Cocoa] Align <datalist> dropdown keyboard interactions with platform conventions
https://bugs.webkit.org/show_bug.cgi?id=286045
rdar://143012287

Reviewed by Abrar Rahman Protyasha.

- Up and down arrow keys should not cycle through options
- Escape key should dismiss the dropdown

* LayoutTests/fast/forms/datalist/datalist-hide-using-escape-key-expected.txt: Added.
* LayoutTests/fast/forms/datalist/datalist-hide-using-escape-key.html: Added.
* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options-on-keydown.html:
* LayoutTests/fast/forms/datalist/datalist-textinput-keydown-expected.txt:
* LayoutTests/fast/forms/datalist/datalist-textinput-keydown.html:
* LayoutTests/fast/forms/datalist/datalist-textinput-suggestions-order.html:
* LayoutTests/platform/gtk/fast/forms/datalist/datalist-textinput-keydown-expected.txt: Added.
* LayoutTests/platform/ipad/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.waitForDataListSuggestionsToChangeVisibility):
(window.UIHelper.isShowingDateTimePicker): Deleted.
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::handleKeydownEvent):
* Source/WebKit/UIProcess/mac/WebDataListSuggestionsDropdownMac.mm:
(-[WKDataListSuggestionsController moveSelectionByDirection:]):
* Source/WebKit/WebProcess/WebCoreSupport/WebDataListSuggestionPicker.cpp:
(WebKit::WebDataListSuggestionPicker::handleKeydownWithIdentifier):

Canonical link: https://commits.webkit.org/289067@main
